### PR TITLE
Add speculative decoding

### PR DESCRIPTION
**Pull request by SamDuffield**
First attempt at adding [speculative decoding](https://arxiv.org/abs/2211.17192)! Discussed in #276 

A few things to note

- It does not support models with different tokenizers, this feels pretty hard to achieve
- It does not currently support vectorised prompt inputs or multiple samples, this probably could be achieved but not sure how desirable since speculative decoding already uses up a lot of parallel resources
- I have tried to make it work for arbitrary `sequence` so including `regex` etc. But in a quite hacky way where I reproduce an instance of the received `sequence`. I think this could be greatly simplified/improved

### CODE REVIEW

**Comment by SamDuffield** [comment]
I realise now that I have coded `parallel_multi_step` in a silly overkill way and all of the pseudo-prompts are not needed! For a transformer you can get logits for the whole previous sequence with just a single call to the model 🤯. Might need to go into the `hidden_state` of the transformer output to extract this and also would need to think about how to do so in a way that conforms to arbitrary outlines `sequences` (regex etc)

**Comment by brandonwillard** [comment]
> * It does not support models with different tokenizers, this feels pretty hard to achieve
> * It does not currently support vectorised prompt inputs or multiple samples, this probably could be achieved but not sure how desirable since speculative decoding already uses up a lot of parallel resources

Yeah, don't worry about these for now.  We can follow up in later PRs.

**Comment by brandonwillard** [comment]
Oh, don't forget to set up `pre-commit` locally!

**Review by brandonwillard — COMMENTED** [review]
OK, so the funnest part of this is going to be testing!  Any ideas for that?

**Comment by brandonwillard** [comment]
Just made a relevant update to the interface in https://github.com/outlines-dev/outlines/pull/318.
